### PR TITLE
Add mandatory upgrade "snode revisions" and start enforcing lokinet/ss versions

### PR DESCRIPTION
**Pull request by jagerman**
Adds a concept of a "snode revision" to the blockchain version.  Snode revisions are a secondary version that let us put out a mandatory update for snodes that isn't a hardfork (and so isn't mandatory for wallets/exchanges/etc.).

Also kills off support for hard fork voting (which apparently was never used, even in Monero land).

Adds minimum lokinet/SS versions to uptime proof acceptance, and puts required versions for proofs to 9.2.0 (oxend), 2.2.0 (storage server) and 0.9.4 (lokinet).  (All of these versions are still in progress, but are the target versions for the 9.2.0 release.)

### CODE REVIEW

File: src/cryptonote_basic/hardfork.cpp [hunk 1]
```diff
@@ -27,468 +26,136 @@
...
-  std::unique_lock l{lock};
-  uint64_t height = db.height();
-  for (auto i = heights.rbegin(); i != heights.rend(); ++i) {
-    if (height >= i->height) {
-      return (i == heights.rbegin() ? i : (i - 1))->version;
-    }
+  std::pair<uint8_t, uint8_t> result;
+  for (auto [it, end] = get_hard_forks(nettype); it != end; it++) {
+    if (it->height <= height)
+      result.first = it->height;
```
Comment (darcys22) [inline src/cryptonote_basic/hardfork.cpp:153]: `result.first = it->version;`

File: src/cryptonote_basic/hardfork.cpp [hunk 2]
```diff
@@ -27,468 +26,136 @@
...
+    result.second = it->height;
```
Comment (darcys22) [inline src/cryptonote_basic/hardfork.cpp:154]: `result.second = it->snode_revision;`

File: src/cryptonote_basic/hardfork.h [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2014-2019, The Monero Project
+// Copyright (c) 2018-2021, The Loki Project
```
Comment (majestrate) [inline src/cryptonote_basic/hardfork.h:1]: oxen?

File: src/cryptonote_basic/hardfork.cpp (hunk 2 above)
Comment (jagerman) [reply src/cryptonote_basic/hardfork.cpp]: Actually it has to be `it->height` because current nodes won't allow a minor version < the major version :(

**Review by darcys22 — APPROVED** [review]

**Pull request merged**